napari/config/core.py

```diff
@@ -0,0 +1,708 @@
+
+        config['_dirty'] = True
+
+    def __enter__(self) -> dict:
+        return self.config
+
+    def __exit__(
+        self,
+        type: Type[Exception],
+        value: Exception,
+        traceback: Optional[TracebackType],
+    ):
+        for op, path, value in reversed(self._record):
+            d = self.config
+            old_val = None
+            if op == "replace":
+                for key in path[:-1]:
+                    d = d.setdefault(key, {})
+                old_val = d[path[-1]]
+                d[path[-1]] = value
+            else:  # insert
+                for key in path[:-1]:
+                    try:
+                        d = d[key]
+                    except KeyError:
+                        break
+                else:
+                    old_val = d.pop(path[-1], None)
+            _callback_to_listeners(path, value, self.callbacks, old_val)
+
+    def _assign(
+        self,
+        keys: Sequence[str],
+        value: Any,
+        d: dict,
+        path: Tuple[str, ...] = (),
+        record: bool = True,
+    ):
+        """Assign value into a nested configuration dictionary
+
+        Parameters
+        ----------
+        keys : Sequence[str]
+            The nested path of keys to assign the value.
+        value : object
+        d : dict
+            The part of the nested dictionary into which we want to assign the
+            value
+        path : Tuple[str], optional
+            The path history up to this point.
+        record : bool, optional
+            Whether this operation needs to be recorded to allow for rollback.
+        """
+        key = canonical_name(keys[0], d)
+
+        path = path + (key,)
+
+        if len(keys) == 1:
+            old_val = None
+            if record:
+                if key in d:
+                    self._record.append(("replace", path, d[key]))
+                    old_val = d[key]
+                else:
+                    self._record.append(("insert", path, None))
+            d[key] = value
+            _callback_to_listeners(path, value, self.callbacks, old_val)
+
+            # might be worth emitting a warning here if we know the value
+            # cannot be serialized
+            # yaml.dump(value, Dumper=ConfigDumper)
+
+        else:
+            if key not in d:
+                if record:
+                    self._record.append(("insert", path, None))
+                d[key] = {}
+                # No need to record subsequent operations after an insert
+                record = False
+            self._assign(keys[1:], value, d[key], path, record=record)
+
+
+def collect_yaml(paths: List[str] = paths) -> List[dict]:
+    """Collect configuration from yaml files.
+
+    This searches through a list of paths, expands to find all yaml or json
+    files, and then parses each file.
+    """
+    logger.debug("collect_yaml from %r", paths)
+
+    # Find all paths
+    file_paths = []
+    for path in paths:
```

Comment:
I see that `paths` variable is not modified here, but mutable object should not be used as default argument

```suggestion
def collect_yaml(paths_list: Optional[List[str]] = None) -> List[dict]:
    """Collect configuration from yaml files.
    This searches through a list of paths, expands to find all yaml or json
    files, and then parses each file.
    """
    if paths_list is None:
        paths_list = paths
    logger.debug("collect_yaml from %r", paths)

    # Find all paths
    file_paths = []
    for path in paths_list: